feat/AB#83067_paginated-refData

matheus-relief opened the pull request:
# Description

This PR adds support to pagination on summary cards with reference data datasource.

On the reference data definition, a few options about pagination were added:
* usePagination: if the refData in question uses pagination
* strategy: 'offset', 'cursor' or 'page'
* other options based on the strategy (those and the strategy itself can be auto-filled based on the query variable names)
On the summary cards:
* selecting reference data as datasource allows the user to map values to the available query variables using the dashboard filters templates
* moving between pages will properly fetch data using our proxy and handles pagination for each strategy


## Useful links

- Please insert link to ticket: https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/83067
- Please insert link to back-end branch if any: https://github.com/ReliefApplications/ems-backend/pull/932

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (refactor or addition to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Reference datas (and api configurations) tested: 

```json
[
  {
    "name": "r&m_episodes",
    "apiConfiguration": {
      "name": "rick_and_morty",
      "status": "active",
      "authType": "public",
      "endpoint": "https://rickandmortyapi.graphcdn.app/",
      "graphQLEndpoint": "/",
      "pingUrl": ""
    },
    "type": "graphql",
    "query": "# You can use {{lastUpdate}} to dynamically get the last fetch date in SQL format.\n\nquery Episodes($page: Int, $episodeName: String) {\n  episodes(page: $page, filter: {name: $episodeName}) {\n    results {\n      id\n      name\n    }\n    info {\n      pages\n      count\n      next\n      prev\n    }\n  }\n}\n",
    "valueField": "id",
    "path": "$.data.episodes.results.*",
    "pageInfo": {
      "strategy": "page",
      "pageVar": "page",
      "totalCountField": "$.data.episodes.info.count"
    }
  },
  {
    "name": "sw_planets",
    "apiConfiguration": {
      "name": "star_wars",
      "status": "active",
      "authType": "public",
      "endpoint": "https://swapi-graphql.netlify.app/.netlify/functions/index/",
      "pingUrl": "/",
      "graphQLEndpoint": "/"
    },
    "type": "graphql",
    "query": "query Planets($cursor: String, $pageSize: Int) {\n  allPlanets(after: $cursor, first: $pageSize) {\n    edges {\n      node {\n        id\n        name\n        diameter\n        gravity\n      }\n      cursor\n    }\n    totalCount\n  }\n}",
    "valueField": "id",
    "path": "$.data.allPlanets.edges.*.node",
    "pageInfo": {
      "strategy": "cursor",
      "cursorField": "$.data.allPlanets.edges.*.cursor",
      "cursorVar": "cursor",
      "pageSizeVar": "pageSize",
      "totalCountField": "$.data.allPlanets.totalCount"
    }
  },
  {
    "name": "pokemon_list",
    "apiConfiguration": {
      "name": "pokemon",
      "status": "active",
      "authType": "public",
      "endpoint": "https://beta.pokeapi.co/graphql/v1beta/",
      "pingUrl": null,
      "graphQLEndpoint": "/"
    },
    "type": "graphql",
    "query": "query Pokemon($pageSize: Int, $offset: Int) {\n  pokemon_v2_pokemonform(limit: $pageSize, offset: $offset) {\n    id\n    name\n  }\n}\n",
    "valueField": "id",
    "path": "$.data.pokemon_v2_pokemonform.*",
    "pageInfo": {
      "strategy": "offset",
      "offsetVar": "offset",
      "pageSizeVar": "pageSize",
      "totalCountField": null
    }
  }
]

```

## Screenshots

![Peek 2023-12-08 00-12](https://github.com/ReliefApplications/ems-frontend/assets/102038450/747ed7db-5dab-4eaa-8833-1a0cdd99a773)


# Checklist:

( * == Mandatory ) 

- [X] * I have set myself as assignee of the pull request
- [X] * My code follows the style guidelines of this project
- [X] * Linting does not generate new warnings
- [X] * I have performed a self-review of my own code
- [X] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [X] * I have commented my code, particularly in hard-to-understand areas
- [X] * I have put JSDoc comment in all required places
- [X] * My changes generate no new warnings
- [X] * I have included screenshots describing my changes if relevant
- [X] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# More explanation
https://www.loom.com/share/05a716d61b9744faaf51fb304c21d1e5?sid=f87cf896-582a-4f76-93ae-8ceed801b145

AntoineRelief commented on apps/back-office/src/app/dashboard/pages/reference-data/graphql/mutations.ts:
```diff
@@ -15,6 +15,7 @@ export const EDIT_REFERENCE_DATA = gql`
     $data: JSON
     $graphQLFilter: String
     $permissions: JSON
+    $pageInfo: PageInfoInput
```
@matheus-relief 
I fear you're going for an incorrect solution
the reference data should stay as it is, no need to modify it I believe

All the logic I had in mind was based on the query definition:

![Capture d’écran 2024-01-11 à 13 01 12](https://github.com/ReliefApplications/ems-frontend/assets/50317271/86c8f2db-1ad0-4b2e-b845-7c32a39a3b43)

it should be kind of acgnostic
you "know" as an admin your query has a "$first" variable, so your widget will pass one to the query, based on current pagination of the widget

AntoineRelief commented:
@matheus-relief 

A few issues I detected so far:

- what to do with the totalCount when API does not provide it? All WHO APIs don't provide a value for that..
- I only saw the monaco editor for graphql variable once, now, when I open the settings, I got an emtpy screen:
![Capture d’écran 2024-01-16 à 22 54 53](https://github.com/ReliefApplications/ems-frontend/assets/50317271/a4b4232d-d1a3-4966-9de6-de976c327b0a)
- I don't think building a getter for availableQueryVariables is a good idea, because you end up with thousand of logs, and I'm sure monaco editor don't want to display based on a getter:

![Capture d’écran 2024-01-16 à 23 05 13](https://github.com/ReliefApplications/ems-frontend/assets/50317271/881fb91b-0164-4512-bcaa-d843d9a08e30)

AntoineRelief commented:
in addition, we don't understand what is the monaco editor, some text to indicate what to do from it could help

AntoineRelief commented:
less priority:
I could not test it, but does it work?:
![Capture d’écran 2024-01-16 à 23 16 53](https://github.com/ReliefApplications/ems-frontend/assets/50317271/8e0e0548-fa3a-4ae9-8efd-73ce3a620bf9)

matheus-relief commented:
@AntoineRelief 

> what to do with the totalCount when API does not provide it? All WHO APIs don't provide a value for that...

If the users doesn't provide the path to the total count or the query fails, we assume that there are infinite items, and hide the page count, and page pagination by number, only allowing users to go to the previous or next pages.

> I only saw the monaco editor for graphql variable once, now, when I open the settings, I got an emtpy screen

The editor only displays when there are variables available (variables defined in the reference data query and not used in the pagination). I added an empty indicator to make it less confusing I hope.

![image](https://github.com/ReliefApplications/ems-frontend/assets/102038450/91f13d30-ce57-4ddd-8748-038ddc5ceaa2)

> I don't think building a getter for availableQueryVariables is a good idea, because you end up with thousand of logs, and I'm sure monaco editor don't want to display based on a getter

Yeah, that was very stupid. Fixed.

> in addition, we don't understand what is the monaco editor, some text to indicate what to do from it could help

I added a tooltip, think it will help. But pretty much is just (some of) the variables that will be sent in the query. In it you can use the same templating syntax we had for the context filters, but differently from it, it's *not* JSON:

```
{
    "region": {{filter.region}}
}
```

Notice that there are no quotes in the template. That allows us to build more complex variables using the filters we have, like in the example bellow

```
{
  "region": {
    "operator": "in",
    "value": [ {{filter.regionA}}, {{filter.regionB}}]
  }
}
```

> I could not test it, but does it work?:

Oh I completely forgot about that. I tested and it doesn't work. Should I fix it or just disable it for now?



# Test with the EMRS_DEV API
![Peek 2023-12-08 00-12](https://github.com/ReliefApplications/ems-frontend/assets/102038450/51c07c3b-7aff-424a-aa72-7cf86b01ffc3)

AntoineRelief commented on libs/shared/src/lib/components/widgets/summary-card/summary-card.component.ts:
```diff
@@ -221,6 +225,46 @@ export class SummaryCardComponent
+
+  /** @returns the graphql query variables object */
+  get graphqlVariables() {
+    const filters = this.contextService.filter.getValue();
```
@matheus-relief 
( for all comments I'll work on it, it's more for indication )
there, I think you reinvent the wheel
the context service has some method for that, so everything there seems to be a duplicated solution

AntoineRelief commented on libs/shared/src/lib/services/reference-data/reference-data.service.ts:
```diff
@@ -151,57 +150,56 @@ export class ReferenceDataService {
+    if (typeof referenceData === 'string') {
+      referenceData = await this.loadReferenceData(referenceData);
+    }
+    const cacheKey = `${referenceData.id || ''}-${JSON.stringify(pageInfo)}`;
```
@matheus-relief 
I think you should not have name it like "pageInfo", because this is not pageInfo but graphql variables
